### ISSUE TITLE
Will JAX support differentiable random variates?

### ISSUE THREAD

**Issue by KristianHolsheimer** [google/jax#2173]
Are there plans to add differentiable variates? For instance, something like:
```python
from jax import random

def normal_differentiable(key, loc, scale):
    # use reparametrization trick
    eps = random.normal(key)
    return loc + eps * scale
```
Of course, for a normal distribution this is kind of trivial, but for something like a categorical distribution it's a little bit more involved (with Gumble-Softmax etc).

It would be great to have these primitives included in `jax.random`.

IMHO this would fit the JAX ethos really well (providing differentiability to previously non-differentiable operations).

**Comment by shoyer** [comment]
I think this is already true for the handful of primitives in `jax.random` with hyper-parameters that output floating point values, e.g., `gamma`, `dirichlet`, `beta`.

I don't know if there are plans to support higher level random variables or probability distributions on top of JAX. I think there is an argument that that is out of scope for the core library. Projects like Tensorflow-Probability already have experimental support for JAX.

**Comment by KristianHolsheimer** [comment]
Ah that's great. I just tried it out for `beta` and in its variates indeed differentiable. Awesome! Thanks @shoyer 

I wonder what would be the choice for having `normal` and `categorical` be non-differentiable. For `categorical` I kind of get it, because it would mean introducing an additional tuning parameter (temperature) to the function signature and the return type would not be a category index (int), but a vector over categories (array<float>).

For `normal`, however, it would be very natural to add the `mu` and `sigma` parameters. In fact, I would go so far as to say that the current state of not being able to pass the distribution parameters to `normal` is inconsistent with the implementations of `gamma`, `dirichlet`, `beta`, which all take distribution parameters.

**Comment by shoyer** [comment]
This is somewhat of a matter of taste, but personally I like writing `mu + sigma * random.normal(key)` more than `random.normal(key, mu, sigma)`. It's about the same number of characters, but is more obvious about what's going on.

I guess the main advantage of adding `loc`/`scale` parameters for normal (whatever they get called) would be that `shape` could be determined automatically based on broadcasting.
